### ISSUE TITLE
TS allow access private class field without error

### ISSUE THREAD

**Issue by terbooter** [Microsoft/TypeScript#10516]
**TypeScript Version:**
```
$ tsc -v
Version 1.8.10
```

Create file app.ts with following content:
```ts
class User {
    constructor(private name:string) {

    }

    public hello(who:User) {
        // name is the private property of object 'who'. TS should show error like
        // "Cat access private field 'name'"
        console.log("Hello, " + who.name);
    }
}

var vasya = new User("VASYA");
var peter = new User("PETER");
vasya.hello(peter);
```
Compile this file
```
tsc app.ts
```
**Actual behavior:** 
You will see no error.

**Expected behavior:**
It should show error
"Cat access private field 'name'"

**Comment by kitsonk** [comment]
BTW, This only occurs when the classes are the same:

```typescript
class Foo {
	constructor(private name: string) {
		
	}
}

class User {
    constructor(private name:string) {

    }

    public hello(who: Foo) {
        console.log("Hello, " + who.name); // Property 'name' is private and is only accessible from within class 'Foo'
    }
}

var vasya = new User("VASYA");
var peter = new Foo("PETER");
vasya.hello(peter);
```

**Comment by RyanCavanaugh** [comment]
Like in most other languages, classes may access private members of other instances of the same class.

**Comment by terbooter** [comment]
@RyanCavanaugh could you please give examples of such languages.
Such behaviour is very surprising for me.

**Comment by terbooter** [comment]
I think `name` field should be `protected` to access it from same class
@ahejlsberg said that 
```
Protected members can be accessed only within the declaring class and subclasses of the declaring class.
```
https://github.com/Microsoft/TypeScript/pull/688

**Comment by kitsonk** [comment]
That would still apply... `User` is the declaring class.  The visibility modifiers are at the class level, not the instance level.

@terbooter there is [PHP](https://secure.php.net/manual/en/language.oop5.visibility.php) and [C#](https://stackoverflow.com/questions/16354728/why-have-class-level-access-modifiers-instead-of-object-level) and Java and pretty much every other OO language.

**Comment by RyanCavanaugh** [comment]
The only language I'm aware of that uses instance-based visibility instead of type-based visibility is Ruby. Java, C#, C++, Swift, PHP, etc. all allow access to other instances' private members.

**Comment by terbooter** [comment]
Thanks for your answers.
